# Hypothesis testing problems

## Exercise 1

You are a nutritionist researching two different types of diets to see if there is a significant difference in weight loss after one month. You choose two random groups of people; one group follows the first diet, and the other follows the second. At the end of the month, you record the weight loss (in kg) for each person:

| Diet 1 | Diet 2 |
|:-------|:-------|
| 2.0 | 3.0 |
| 2.5 | 3.2 |
| 3.0 | 3.1 |
| 2.8 | 2.9 |
| 2.3 | 2.8 |
| 2.7 | 3.0 |
| 2.5 | 3.2 |

With these data, it seeks to answer the following question: Is there a significant difference in average weight loss between people who followed the first diet and those who followed the second diet?

To draw conclusions, follow the points below:

- State the hypothesis: null and alternative hypothesis.
- Perform the test to test the hypothesis. You can use a Student's t-test.
- Analyze the conclusions.

In [2]:
# TODO
# Null Hypothesys: The average weight loss between people who followed the first diet is equal to those who followed the second diet.
# h0 = d1.mean = d2.mean
# Alternative Hypothesys: There is a significant different between average weight loss between people who followed diet1 vs who followed diet2.
# h1 = d1.mean != d2.mean

In [2]:
import pandas as pd
import numpy as np
import scipy.stats as stats

In [3]:
diet1 = [2.0, 2.5, 3.0, 2.8, 2.3, 2.7, 2.5]
diet2 = [3.0, 3.2, 3.1, 2.9, 2.8, 3, 3.2]
print(np.array(diet1).mean())
print(np.array(diet2).mean())


2.542857142857143
3.0285714285714285


In [4]:
# Sample standard deviation
diet1_std = np.array(diet1).std()
diet2_std = np.array(diet2).std()
print(f" Diet1 std: {diet1_std}\nDiet2 std: {diet2_std}")

 Diet1 std: 0.306394436993246
Diet2 std: 0.13850513878332382


In [5]:
_, p_value = stats.ttest_ind(diet1, diet2)
p_value

np.float64(0.004083270191713912)

In [6]:
sigma = 0.05
if p_value < sigma:
    print(F"Rechazamos H0, porque la media muestral sugieren que la diet2 es significativamente diferente a diet1")
else:
    print(F"Aceptamos H0, porque no hay suficientes evidencias para demostrar diferencias entre la diet1 y diet2")

Rechazamos H0, porque la media muestral sugieren que la diet2 es significativamente diferente a diet1


## ANOVA

**ANOVA** (*Analysis of Variance*) is a statistical technique used to compare the measures of two or more groups. The idea behind ANOVA is to decompose the total variability in the data into two components: between-group variability and within-group variability:

- **Between-group variability**: This variability refers to the differences between the group means. If this variability is considerably larger than the within-group variability, it could be an indication that at least one of the group means is different.
- **Within-group variability**: This variability refers to the dispersion of the data within each group. If all groups have similar variability, then any noticeable difference in group means could be considered significant.

Hypotheses in ANOVA typically include:

- **Null hypothesis** ($H₀$): The means of all groups are equal.
- **Alternative hypothesis** ($H₁$): At least one of the group means is different.

If the ANOVA test result is significant (e.g., a p-value less than a threshold such as 0.05), this suggests that at least one group mean is different.

## Exercise 2

A farmer decides to test three different types of fertilizers to determine if one is superior in terms of corn production. The farmer plants corn on 15 identical plots and uses all three fertilizers (5 plots for each type). At the end of the season, he measures the corn yield (in kg) of each plot, with the following result:

| Fertilizer 1 | Fertilizer 2 | Fertilizer 3 |
|:-------------|:-------------|:-------------|
| 20 | 22 | 24 |
| 21 | 21 | 23 |
| 20 | 23 | 22 |
| 19 | 22 | 23 |
| 20 | 21 | 24 |

With this data, he seeks to answer the following question: Is there a significant difference in average corn yield between the three types of fertilizers?

To help you, follow the points below:

- State the hypothesis: null and alternative hypothesis.
- Perform the ANOVA test.
- Analyze the conclusions.
- If one fertilizer is better than another, how can we know it?

In [7]:
# Null Hypothesys: The average corn for each yield in kg will be the same for all three types of fertilizers
# h0 = f1.mean = f2.mean = f3.mean
# Alternative Hypothesys: There is a significant different between average in one corn yield between the three fertilizers
# h1 = f1.mean != f2.mean != f3.mean

In [8]:
f1 = [20, 21, 20, 19, 20]
f2 = [22, 21, 23, 22, 21]
f3 = [24, 23, 22, 23, 24]

In [9]:
_, p_value = stats.f_oneway(f1, f2, f3)
sigma = 0.05
if p_value < sigma:
    print(F"Rechazamos H0, porque la media muestral sugieren que existe una gran diferencia en los resultados medios de corn para los 3 fertilizantes")
else:
    print(F"Aceptamos H0, porque no hay suficientes evidencias para demostrar una gran diferencia en la media de kg para cada fertilizantes en los corn")



Rechazamos H0, porque la media muestral sugieren que existe una gran diferencia en los resultados medios de corn para los 3 fertilizantes


In [14]:
from statsmodels.stats.multicomp import pairwise_tukeyhsd
yield_data = np.array(f1 + f2 + f3)
fertilizer_group = (['F1'] * len(f1) + ['F2'] * len(f2) + ['F3'] * len(f3))

data_long = pd.DataFrame({
    'Yield': yield_data,
    'Fertilizer': fertilizer_group
})
tukey_results = pairwise_tukeyhsd(endog=data_long['Yield'], 
                                  groups=data_long['Fertilizer'],
                                  alpha=0.05)

print("Tukey HSD Post-Hoc Test Results:")
print(tukey_results)

Tukey HSD Post-Hoc Test Results:
Multiple Comparison of Means - Tukey HSD, FWER=0.05
group1 group2 meandiff p-adj  lower  upper  reject
--------------------------------------------------
    F1     F2      1.8 0.0099 0.4572 3.1428   True
    F1     F3      3.2 0.0001 1.8572 4.5428   True
    F2     F3      1.4 0.0409 0.0572 2.7428   True
--------------------------------------------------


# Conclusion from the tukey test table
- Fertilizer 2 is better than F1 where difference is 1.8 kg. a p_value 0.0099 
- F3 is much better than F1. The difference is 3.2 kg. and highly significant in the comparison
- Betweeen F2 vs F3 the difference is 1.4 kg. very significant aswell making F3 much better
- We can clode that F3 represents the highest diference between the three fertilizers. 